File: ScaleformUI_Csharp/Menus/RadialMenu/RadialMenu.cs
using CitizenFX.Core;
using CitizenFX.FiveM;
using CitizenFX.FiveM.GUI;
using ScaleformUI.Elements;
using ScaleformUI.Menu;
using ScaleformUI.Menus;
using ScaleformUI.Scaleforms;
using System.Drawing;
using static CitizenFX.FiveM.Native.Natives;

namespace ScaleformUI.Radial
{
    public delegate void MenuOpenedEvent(RadialMenu menu, dynamic data = null);
    public delegate void MenuClosedEvent(RadialMenu menu);
    public delegate void SegmentChanged(RadialSegment segment);
    public delegate void IndexChanged(RadialSegment segment, int index);
    public delegate void SegmentSelected(RadialSegment segment);

    public class RadialMenu : MenuBase
    {
        private bool visible;
        private int currentSelection;
        private float oldAngle;
        private bool changed;
        public RadialSegment[] Segments { get; private set; }
        private PointF _offset = new PointF(0, 0);

        public MenuOpenedEvent OnMenuOpen;
        public MenuClosedEvent OnMenuClose;
        public SegmentChanged OnSegmentHighlight;
        public IndexChanged OnSegmentIndexChange;
        public SegmentSelected OnSegmentSelect;
        private bool enable3D = true;

        public bool Enable3D
        {
            get => enable3D;
            set
            {
                enable3D = value;
                if (Visible)
                {
                    Main.radialMenu.CallFunction("ENABLE_3D", value);
                }
            }
        }

        public override bool Visible
        {
            get => visible;
            set
            {
                visible = value;
                if (value)
                {
                    BuildMenu();
                    MenuHandler.currentMenu = this;
                    MenuHandler.ableToDraw = true;
                    Main.InstructionalButtons.SetInstructionalButtons(InstructionalButtons);
                    MenuOpenEv(this, null);
                    if (BreadcrumbsHandler.Count == 0)
                        BreadcrumbsHandler.Forward(this, null);
                }
                else
                {
                    Main.InstructionalButtons.ClearButtonList();
                    MenuHandler.ableToDraw = false;
                    MenuHandler.currentMenu = null;
                    Main.radialMenu.CallFunction("CLEAR_ALL");
                    MenuCloseEv(this);
                }
            }
        }

        public RadialMenu() : this(default) { }

        public RadialMenu(PointF offset)
        {
            _offset = offset;
            Segments = new RadialSegment[8]
            {
                new RadialSegment(0, this),
                new RadialSegment(1, this),
                new RadialSegment(2, this),
                new RadialSegment(3, this),
                new RadialSegment(4, this),
                new RadialSegment(5, this),
                new RadialSegment(6, this),
                new RadialSegment(7, this),
            };
            InstructionalButtons = new List<InstructionalButton>()
            {
                new InstructionalButton(Control.PhoneSelect, UIMenu._selectTextLocalized),
                new InstructionalButton(Control.PhoneCancel, UIMenu._backTextLocalized)
            };
        }

        public void AddInstructionalButton(InstructionalButton button)
        {
            InstructionalButtons.Add(button);
            if (Visible && !(Main.Warning.IsShowing || Main.Warning.IsShowingWithButtons))
                Main.InstructionalButtons.SetInstructionalButtons(InstructionalButtons);
        }

        public void RemoveInstructionalButton(InstructionalButton button)
        {
            if (InstructionalButtons.Contains(button))
                InstructionalButtons.Remove(button);
            if (Visible && !(Main.Warning.IsShowing || Main.Warning.IsShowingWithButtons))
                Main.InstructionalButtons.SetInstructionalButtons(InstructionalButtons);
        }

        public void RemoveInstructionalButton(int index)
        {
            if (index < 0 || index >= InstructionalButtons.Count)
                throw new ArgumentOutOfRangeException("ScaleformUI: Cannot remove with an index less than 0 or more than the count of actual instructional buttons");
            RemoveInstructionalButton(InstructionalButtons[index]);
        }

        internal async void BuildMenu()
        {
            if (_offset.IsEmpty) _offset = new PointF(Screen.Width / 2, (Screen.Height / 2) - 60);
            Main.radialMenu.CallFunction("CREATE_MENU", enable3D, _offset.X, _offset.Y);
            for (int i = 0; i < 8; i++)
            {
                RadialSegment segment = Segments[i];
                for (int j = 0; j < segment.Items.Count; j++)
                {
                    SegmentItem item = segment.Items[j];
                    Main.radialMenu.CallFunction("ADD_ITEM", i, item.Label, item.Description, item.TextureDict, item.TextureName, item.TextureWidth, item.TextureHeight, item.Color, item.qtty, item.max);
                }
            }
            Main.radialMenu.CallFunction("LOAD_MENU", currentSelection, Segments[0].CurrentSelection, Segments[1].CurrentSelection, Segments[2].CurrentSelection, Segments[3].CurrentSelection, Segments[4].CurrentSelection, Segments[5].CurrentSelection, Segments[6].CurrentSelection, Segments[7].CurrentSelection);
        }

        internal override async void ProcessControl(Keys key = Keys.None)
        {
            Controls.Toggle(false);
            // block camera movements
            Game.DisableControlThisFrame(0, Control.LookLeftRight);
            Game.DisableControlThisFrame(0, Control.LookUpDown);

            // take mouse/gamepad LStick
            float x = (float)Math.Floor(GetDisabledControlNormal(2, 13) * 1000);
            float y = (float)Math.Floor(GetDisabledControlNormal(2, 12) * 1000);

            // math.atan2 returns 0 when mouse/gamepad LStick move right, because y=0 and x is not negative..
            // as a workaround.. i set y = 1, this way left is checked correctly
            if (x > 0 && y == 0) y = 1;
            float angle = 0;
            float normalized_angle = 0;
            int finalizedWorkingAngle = -1;

            if (x > 400 || y > 400 || x < -400 || y < -400)
            {

                // we convert controls into angle degrees (rad) and then into -180° -> 180°
                angle = (float)(Math.Atan2(y, x) * (180 / Math.PI));
                // normalize the angle into 0-360° range
                normalized_angle = (float)Math.Floor(angle != 0 ? (angle + 450) % 360 : 0);

                // then we get its value between 0-7.. or -1 if nothing has been pressed.
                finalizedWorkingAngle = (int)Math.Floor(angle != 0 ? normalized_angle / 45 : -1);

                if (currentSelection != finalizedWorkingAngle && finalizedWorkingAngle != -1 && (normalized_angle > oldAngle + 45 || normalized_angle < oldAngle - 45))
                {
                    if (!changed)
                    {
                        Segments[currentSelection].Selected = false;
                        currentSelection = finalizedWorkingAngle;
                        oldAngle = normalized_angle;
                        changed = true;
                    }
                }

                if (changed)
                {
                    Segments[currentSelection].Selected = true;
                    OnSegmentHighlight?.Invoke(Segments[currentSelection]);
                    Main.radialMenu.CallFunction("SET_POINTER", finalizedWorkingAngle, true);
                    changed = false;
                }
            }

            if (Game.IsControlJustPressed(0, Control.WeaponWheelPrev))
            {
                int sel = await Segments[currentSelection].CycleItems(-1);
                OnSegmentIndexChange?.Invoke(Segments[currentSelection], sel);
            }

            if (Game.IsControlJustPressed(0, Control.WeaponWheelNext))
            {
                int sel = await Segments[currentSelection].CycleItems(1);
                OnSegmentIndexChange?.Invoke(Segments[currentSelection], sel);
            }

            if (Game.IsControlJustPressed(0, Control.FrontendCancel))
            {
                GoBack();
            }

            if (Game.IsControlJustPressed(0, Control.FrontendAccept))
            {
                Select();
            }
        }

        internal void GoBack()
        {
            if (BreadcrumbsHandler.CurrentDepth == 0)
            {
                Visible = false;
                BreadcrumbsHandler.Clear();
                Main.InstructionalButtons.ClearButtonList();
            }
            else
            {
                BreadcrumbsHandler.SwitchInProgress = true;
                MenuBase prevMenu = BreadcrumbsHandler.PreviousMenu;
                BreadcrumbsHandler.Backwards();
                Visible = false;
                prevMenu.Visible = true;
                BreadcrumbsHandler.SwitchInProgress = false;
            }
        }

        internal void Select()
        {
            OnSegmentSelect?.Invoke(Segments[currentSelection]);
        }

        internal override void Draw()
        {
            Screen.Hud.HideComponentThisFrame(HudComponent.WeaponWheel);
            Main.radialMenu.Render2D();
        }

        public int CurrentSegment
        {
            get => currentSelection;
            set
            {
                currentSelection = value;
                if (Visible)
                {
                    Main.radialMenu.CallFunction("SET_POINTER", value, true);
                }
            }
        }

        internal void MenuOpenEv(RadialMenu menu, dynamic data)
        {
            OnMenuOpen?.Invoke(menu, data);
        }

        internal void MenuCloseEv(RadialMenu menu)
        {
            OnMenuClose?.Invoke(menu);
        }
    }
}

File: ScaleformUI_Lua/src/Menus/RadialMenu/RadialMenu.lua
RadialMenu = setmetatable({}, RadialMenu)
RadialMenu.__index = RadialMenu
RadialMenu.__call = function()
    return "RadialMenu"
end

---@class RadialMenu: Scaleform
---@field public Segments table
---@field public InstructionalButtons table
---@field public OnMenuOpen fun(menu:RadialMenu, data:any)
---@field public OnMenuClose fun(menu:RadialMenu)
---@field public OnSegmentHighlight fun(segment:SegmentItem)
---@field public OnSegmentIndexChange fun(segment:SegmentItem, index:number)
---@field public OnSegmentSelect fun(segment:SegmentItem)
---@field public currentSelection number
---@field public oldAngle number
---@field public changed boolean
---@field public enable3D boolean
---@field public offset table
---@field public AddInstructionButton fun(button:InstructionalButton)
---@field public RemoveInstructionButton fun(button:table)
---@field public Enable3D fun(enable:boolean)
---@field public CurrentSelection fun(index:number)
---@field public Visible fun(bool:boolean)
---@field public BuildMenu fun()
---@field public ProcessMouse fun()
---@field public ProcessControl fun()
---@field public SwitchTo fun(newMenu:RadialMenu, newMenuCurrentSelection:number, inheritOldMenuParams:boolean)
---@field public GoBack fun()
---@field public Select fun()
---@field public Draw fun()
---@field public New fun(x:number, y:number):RadialMenu

---New
---@param x number
---@param y number
---@return RadialMenu
function RadialMenu.New(x, y)
    local X, Y = tonumber(x) or 0, tonumber(y) or 0
    local _rad = {
        _visible = false,
        currentSelection = 1,
        oldAngle = 0,
        changed = false,
        enable3D = true,
        offset = { x = X, y = Y },
        Segments = {
            RadialSegment.New(1),
            RadialSegment.New(2),
            RadialSegment.New(3),
            RadialSegment.New(4),
            RadialSegment.New(5),
            RadialSegment.New(6),
            RadialSegment.New(7),
            RadialSegment.New(8),
        },
        InstructionalButtons = {
            InstructionalButton.New(GetLabelText("HUD_INPUT2"), -1, 176, 176, -1),
            InstructionalButton.New(GetLabelText("HUD_INPUT3"), -1, 177, 177, -1)
        },
        OnMenuOpen = function(menu, data)
        end,
        OnMenuClose = function(menu)
        end,
        OnSegmentHighlight = function(segment)
        end,
        OnSegmentIndexChange = function(segment, index)
        end,
        OnSegmentSelect = function(segment)
        end
    }
    local meta = setmetatable(_rad, RadialMenu)
    meta.Segments[1].Parent = meta
    meta.Segments[2].Parent = meta
    meta.Segments[3].Parent = meta
    meta.Segments[4].Parent = meta
    meta.Segments[5].Parent = meta
    meta.Segments[6].Parent = meta
    meta.Segments[7].Parent = meta
    meta.Segments[8].Parent = meta
    return meta
end

---AddInstructionButton
---@param button InstructionalButton
function RadialMenu:AddInstructionButton(button)
    if type(button) == "table" then
        self.InstructionalButtons[#self.InstructionalButtons + 1] = button
        if self:Visible() and not ScaleformUI.Scaleforms.Warning:IsShowing() then
            ScaleformUI.Scaleforms.InstructionalButtons:SetInstructionalButtons(self.InstructionalButtons)
        end
    end
end

---RemoveInstructionButton
---@param button table
function RadialMenu:RemoveInstructionButton(button)
    if type(button) == "table" then
        for i = 1, #self.InstructionalButtons do
            if button == self.InstructionalButtons[i] then
                table.remove(self.InstructionalButtons, i)
                break
            end
        end
    else
        if tonumber(button) then
            if self.InstructionalButtons[tonumber(button)] then
                table.remove(self.InstructionalButtons, tonumber(button))
            end
        end
    end
    if self:Visible() and not ScaleformUI.Scaleforms.Warning:IsShowing() then
        ScaleformUI.Scaleforms.InstructionalButtons:SetInstructionalButtons(self.InstructionalButtons)
    end
end

function RadialMenu:Enable3D(enable)
    if enable ~= nil then
        self.enable3D = enable
        if self:Visible() then
            ScaleformUI.Scaleforms._radialMenu:CallFunction("ENABLE_3D", enable)
        end
    else
        return self.enable3D
    end
end

function RadialMenu:CurrentSelection(index)
    if index ~= nil then
        self.currentSelection = index
        if self:Visible() then
            ScaleformUI.Scaleforms._radialMenu:CallFunction("SET_POINTER", (index - 1), true)
        end
    else
        return self.currentSelection or 1
    end
end

function RadialMenu:Visible(bool)
    if bool ~= nil then
        self._visible = bool
        if bool then
            self:BuildMenu()
            MenuHandler._currentMenu = self
            MenuHandler.ableToDraw = true
            self.OnMenuOpen(self)
            if BreadcrumbsHandler:Count() == 0 then
                BreadcrumbsHandler:Forward(self)
            end
        else
            ScaleformUI.Scaleforms.InstructionalButtons:ClearButtonList()
            self.OnMenuClose(self)
            ScaleformUI.Scaleforms._radialMenu:CallFunction("CLEAR_ALL")
            MenuHandler.ableToDraw = false
        end
    else
        return self._visible
    end
end

function RadialMenu:BuildMenu()
    ScaleformUI.Scaleforms._radialMenu:CallFunction("CREATE_MENU", self:Enable3D(), (1280 / 2) + self.offset.x, ((720 / 2) - 60) + self.offset.y)
    for i = 1, 8 do
        local seg = self.Segments[i]
        for j = 1, #seg.Items do
            local item = seg.Items[j]
            ScaleformUI.Scaleforms._radialMenu:CallFunction("ADD_ITEM", i - 1, item:Label(), item:Description(), item:TextureDict(), item:TextureName(), item:TextureWidth(), item:TextureHeight(), item:Color(), item.qtty, item.max)
        end
    end
    ScaleformUI.Scaleforms._radialMenu:CallFunction("LOAD_MENU", self.currentSelection - 1, self.Segments[1]:CurrentSelection() - 1, self.Segments[2]:CurrentSelection() - 1, self.Segments[3]:CurrentSelection() - 1, self.Segments[4]:CurrentSelection() - 1, self.Segments[5]:CurrentSelection() - 1, self.Segments[6]:CurrentSelection() - 1, self.Segments[7]:CurrentSelection() - 1, self.Segments[8]:CurrentSelection() - 1)
end

function RadialMenu:ProcessMouse()
end

function RadialMenu:ProcessControl()
    if not self:Visible() then
        return
    end
    if UpdateOnscreenKeyboard() == 0 or IsWarningMessageActive() or ScaleformUI.Scaleforms.Warning:IsShowing() or BreadcrumbsHandler.SwitchInProgress then return end

    Controls:ToggleAll(false)
    DisableControlAction(0, 1, true)
    DisableControlAction(0, 2, true)

    local x = math.floor(GetDisabledControlNormal(2, 13) * 1000)
    local y = math.floor(GetDisabledControlNormal(2, 12) * 1000)

    if x > 0 and y == 0 then y = 1 end
    local angle = 0
    local normalized_angle = 0
    local finalizedAngle = -1
    if x > 400 or y > 400 or x < -400 or y < -400 then
        angle = math.atan(y, x) * (180 / math.pi)
        if angle == 0 then
            normalized_angle = 0
        else
            normalized_angle = (angle + 450) % 360
        end
        if angle == 0 then
            finalizedAngle = -1
        else
            finalizedAngle = math.floor(normalized_angle / 45)
        end

        if self.currentSelection ~= finalizedAngle + 1 and finalizedAngle ~= -1 and (normalized_angle > self.oldAngle + 45 or normalized_angle < self.oldAngle - 45) then
            if not self.changed then
                self.Segments[self.currentSelection].Selected = false
                self.currentSelection = finalizedAngle + 1
                self.oldAngle = normalized_angle
                self.changed = true
            end
        end

        if self.changed then
            self.Segments[self.currentSelection].Selected = true
            self.OnSegmentHighlight(self.Segments[self.currentSelection])
            ScaleformUI.Scaleforms._radialMenu:CallFunction("SET_POINTER", finalizedAngle, true)
            self.changed = false
        end
    end

    if IsDisabledControlJustPressed(0, 15) then
        Citizen.CreateThread(function()
            local sel = self.Segments[self.currentSelection]:CycleItems(-1)
            self.OnSegmentIndexChange(self.Segments[self.currentSelection], sel)
        end)
    end

    if IsDisabledControlJustPressed(0, 14) then
        Citizen.CreateThread(function()
            local sel = self.Segments[self.currentSelection]:CycleItems(1)
            self.OnSegmentIndexChange(self.Segments[self.currentSelection], sel)
        end)
    end

    if IsDisabledControlJustPressed(0, 202) then
        self:GoBack()
    end

    if IsDisabledControlJustPressed(0, 201) then
        self:Select()
    end
end

function RadialMenu:SwitchTo(newMenu, newMenuCurrentSelection, inheritOldMenuParams)
    MenuHandler:SwitchTo(self, newMenu, newMenuCurrentSelection, inheritOldMenuParams)
end

function RadialMenu:GoBack()
    if BreadcrumbsHandler:CurrentDepth() == 1 then
        self:Visible(false)
        BreadcrumbsHandler:Clear()
    else
        BreadcrumbsHandler.SwitchInProgress = true
        local prevMenu = BreadcrumbsHandler:PreviousMenu()
        BreadcrumbsHandler:Backwards()
        self:Visible(false)
        prevMenu.menu:Visible(true)
        BreadcrumbsHandler.SwitchInProgress = false
    end
end

function RadialMenu:Select()
    self.OnSegmentSelect(self.Segments[self.currentSelection])
end

function RadialMenu:Draw()
    HideHudComponentThisFrame(19)
    ScaleformUI.Scaleforms._radialMenu:Render2D()
end
